network/tcp.go

```diff
@@ -0,0 +1,395 @@
+	// Stop() have been made (concurrency into place here)
+	closed bool
+
+	// listening addr (actual). Useful for listening on :0 port
+	addr net.Addr
+}
+
+// NewTCPListener returns a Listener. This function tries to bind to the given
+// address already.It returns the listener and an error if one occured during
+// the binding. A subsequent call to Address() will give the real listening
+// address (useful if you set it to port :0).
+func NewTCPListener(addr Address) (*TCPListener, error) {
+	t := &TCPListener{
+		quit:         make(chan bool),
+		quitListener: make(chan bool),
+	}
+	global, _ := GlobalBind(addr.NetworkAddress())
+	for i := 0; i < MaxRetryConnect; i++ {
+		ln, err := net.Listen("tcp", global)
+		if err == nil {
+			t.listener = ln
+			break
+		} else if i == MaxRetryConnect-1 {
+			return nil, errors.New("Error opening listener: " + err.Error())
+		}
+		time.Sleep(WaitRetry)
+	}
+	t.addr = t.listener.Addr()
+	return t, nil
+}
+
+// Listen implements the Listener interface
+func (t *TCPListener) Listen(fn func(Conn)) error {
+	receiver := func(tc *TCPConn) {
+		go fn(tc)
+	}
+	return t.listen(receiver)
+}
+
+// listen is the private function that takes a function that takes a TCPConn.
+// That way we can control what to do of the TCPConn before returning it to the
+// function given by the user. fn is called in the same routine.
+func (t *TCPListener) listen(fn func(*TCPConn)) error {
+	t.listeningLock.Lock()
+	if t.closed == true {
+		t.listeningLock.Unlock()
+		return nil
+	}
+	t.listening = true
+	t.listeningLock.Unlock()
+	for {
+		conn, err := t.listener.Accept()
+		if err != nil {
+			select {
+			case <-t.quit:
+				t.quitListener <- true
+				return nil
+			default:
+			}
+			continue
+		}
+		c := TCPConn{
+			endpoint: NewTCPAddress(conn.RemoteAddr().String()),
+			conn:     conn,
+		}
+		fn(&c)
+	}
+}
+
+// Stop will stop the listener. It is a blocking call.
+func (t *TCPListener) Stop() error {
+	// lets see if we launched a listening routing
+	t.listeningLock.Lock()
+	defer t.listeningLock.Unlock()
+
+	close(t.quit)
+
+	if t.listener != nil {
+		if err := t.listener.Close(); err != nil {
+			if handleError(err) != ErrClosed {
+				return err
+			}
+		}
+	}
+	var stop bool
+	if t.listening {
+		for !stop {
+			select {
+			case <-t.quitListener:
+				stop = true
+			case <-time.After(time.Millisecond * 50):
+				continue
+			}
+		}
+	}
+
+	t.quit = make(chan bool)
+	t.listening = false
+	t.closed = true
+	return nil
+}
+
+// Address implements the Listener interface
+func (t *TCPListener) Address() Address {
+	t.listeningLock.Lock()
+	defer t.listeningLock.Unlock()
+	return NewAddress(PlainTCP, t.addr.String())
+}
+
+// Listening implements the Listener interface
+func (t *TCPListener) Listening() bool {
+	t.listeningLock.Lock()
+	defer t.listeningLock.Unlock()
+	return t.listening
+}
+
+// TCPHost implements the Host interface using TCP connections.
+type TCPHost struct {
+	addr Address
```

Comment:
wow you're usually the one with the short variable. Golang also is a fan of short variable and I find `addr` pretty specific in that context ...